components/brave_rewards/resources/extension/brave_rewards/background/publisherVisit.ts

```diff
@@ -0,0 +1,240 @@
+
+let publisherKeys: PublisherKeyDictionary = {}
+
+const handleGreaselionErrorResponse = (tabId: number, mediaType: string, data: GreaselionErrorResponse) => {
+  console.error(`Greaselion error: ${data.errorMessage}`)
+}
+
+const handleMediaDurationMetadataResponse = (tabId: number, mediaType: string, data: MediaDurationMetadataResponse) => {
+  const publisherKey = publisherKeys[data.mediaKey]
+  if (!publisherKey) {
+    console.error(`Failed to handle media duration metadata response: missing publisher key for media key ${data.mediaKey}`)
+    return
+  }
+
+  chrome.braveRewards.getPublisherInfo(publisherKey, (result: number, info?: RewardsExtension.Publisher) => {
+    console.debug(`getPublisherInfo: result=${result}`)
+
+    if (result === 0 && info) {
+      console.info('Updating media duration:')
+      console.info(`  tabId=${tabId}`)
+      console.info(`  publisherKey=${info.publisher_key}`)
+      console.info(`  publisherName=${info.name}`)
+      console.info(`  duration=${data.duration}`)
+
+      chrome.braveRewards.updateMediaDuration(
+        tabId,
+        mediaType,
+        publisherKey,
+        data.duration)
+      return
+    }
+  })
+}
+
+const handleMediaDurationHandlerRegistrationRequest = (tabId: number, mediaType: string, data: MediaDurationHandlerRegistrationRequest) => {
+  // If we already registered a handler for this tab, exit early
+  if (mediaDurationHandlerRegistrations[tabId] && mediaDurationHandlerRegistrations[tabId].registered) {
+    return
+  }
+
+  // Mark this tab as registered
+  mediaDurationHandlerRegistrations[tabId] = {
+    registered: true
+  }
+
+  chrome.webRequest.onCompleted.addListener(
+    // Listener
+    function (details) {
+      const port = greaselionPorts[tabId]
+      if (!port) {
+        return
+      }
+      port.postMessage({
+        msgType: 'MediaDurationMetadataRequest',
+        url: details.url
+      })
+    },
+    // Filters
+    {
+      types: [
+        'image',
+        'media',
+        'script',
+        'xmlhttprequest'
+      ],
+      urls: [
+        data.urlRegex
+      ]
+    })
+}
+
+const handleSavePublisherVisitResponse = (tabId: number, mediaType: string, data: SavePublisherVisitResponse) => {
+  console.info('Visited a publisher url:')
+  console.info(`  tabId=${tabId}`)
+  console.info(`  mediaType=${mediaType}`)
+  console.info(`  url=${data.url}`)
+  console.info(`  publisherKey=${data.publisherKey}`)
+  console.info(`  publisherName=${data.publisherName}`)
+  console.info(`  favIconUrl=${data.favIconUrl}`)
+  console.info(`  mediaKey=${data.mediaKey}`)
+
+  if (!data.publisherKey) {
+    console.error('Failed to handle publisher visit: missing publisher key')
+    return
+  }
+
+  if (data.mediaKey && !publisherKeys[data.mediaKey]) {
+    publisherKeys[data.mediaKey] = data.publisherKey
+  }
+
+  chrome.braveRewards.getPublisherInfo(data.publisherKey, (result: number, info?: RewardsExtension.Publisher) => {
```

Comment:
Similar to the other call to `getPublisherInfo`, I'm wondering if we can just roll the `result` checking behavior into `savePublisherVisit`?

Reply:
I'm going to hold off on this one for now, as it's not a completely straightforward refactoring. I do think it makes sense though, I'll create an issue to track it.